File: ShellAI/java/BaseAI.java
import com.sun.jna.Pointer;

/// \brief A basic AI interface.

///This class implements most the code an AI would need to interface with the lower-level game code.
///AIs should extend this class to get a lot of builer-plate code out of the way
///The provided AI class does just that.
public abstract class BaseAI
{
  static Player[] players;
  static Mappable[] mappables;
  static Plant[] plants;
  static Mutation[] mutations;
  Pointer connection;
  static int iteration;
  boolean initialized;

  public BaseAI(Pointer c)
  {
    connection = c;
  }
    
  ///
  ///Make this your username, which should be provided.
  public abstract String username();
  ///
  ///Make this your password, which should be provided.
  public abstract String password();
  ///
  ///This is run on turn 1 before run
  public abstract void init();
  ///
  ///This is run every turn . Return true to end the turn, return false
  ///to request a status update from the server and then immediately rerun this function with the
  ///latest game status.
  public abstract boolean run();

  ///
  ///This is run on after your last turn.
  public abstract void end();


  public boolean startTurn()
  {
    iteration++;
    int count = 0;
    count = Client.INSTANCE.getPlayerCount(connection);
    players = new Player[count];
    for(int i = 0; i < count; i++)
    {
      players[i] = new Player(Client.INSTANCE.getPlayer(connection, i));
    }
    count = Client.INSTANCE.getMappableCount(connection);
    mappables = new Mappable[count];
    for(int i = 0; i < count; i++)
    {
      mappables[i] = new Mappable(Client.INSTANCE.getMappable(connection, i));
    }
    count = Client.INSTANCE.getPlantCount(connection);
    plants = new Plant[count];
    for(int i = 0; i < count; i++)
    {
      plants[i] = new Plant(Client.INSTANCE.getPlant(connection, i));
    }
    count = Client.INSTANCE.getMutationCount(connection);
    mutations = new Mutation[count];
    for(int i = 0; i < count; i++)
    {
      mutations[i] = new Mutation(Client.INSTANCE.getMutation(connection, i));
    }

    if(!initialized)
    {
      initialized = true;
      init();
    }
    return run();
  }


  ///The width of the total map.
  int mapWidth()
  {
    return Client.INSTANCE.getMapWidth(connection);
  }
  ///The height of the total map.
  int mapHeight()
  {
    return Client.INSTANCE.getMapHeight(connection);
  }
  ///The current turn number.
  int turnNumber()
  {
    return Client.INSTANCE.getTurnNumber(connection);
  }
  ///The maximum number of Plants allowed per player.
  int maxPlants()
  {
    return Client.INSTANCE.getMaxPlants(connection);
  }
  ///The id of the current player.
  int playerID()
  {
    return Client.INSTANCE.getPlayerID(connection);
  }
  ///What number game this is for the server.
  int gameNumber()
  {
    return Client.INSTANCE.getGameNumber(connection);
  }
  ///The maximum distance that a bumbleweed can move during an uproot
  int bumbleweedSpeed()
  {
    return Client.INSTANCE.getBumbleweedSpeed(connection);
  }
  ///The damage taken from standing in a pool
  int poolDamage()
  {
    return Client.INSTANCE.getPoolDamage(connection);
  }
  ///The strength buff taken from standing in a pool
  int poolBuff()
  {
    return Client.INSTANCE.getPoolBuff(connection);
  }
  ///The rate which your spores return to maxSpores
  int sporeRate()
  {
    return Client.INSTANCE.getSporeRate(connection);
  }
  ///The maximum number of spores a player can have
  int maxSpores()
  {
    return Client.INSTANCE.getMaxSpores(connection);
  }
  ///The uproot range
  int uprootRange()
  {
    return Client.INSTANCE.getUprootRange(connection);
  }
}

File: ShellAI/csharp/BaseAI.cs
using System;
using System.Runtime.InteropServices;

/// <summary>
/// This class implements most the code an AI would need to interface with the lower-level game code.
/// AIs should extend this class to get a lot of builer-plate code out of the way. The provided AI class does just that.
/// </summary>
public abstract class BaseAI
{
  public static Player[] players;
  public static Mappable[] mappables;
  public static Plant[] plants;
  public static Mutation[] mutations;

  IntPtr connection;
  public static int iteration;
  bool initialized;

  public BaseAI(IntPtr c)
  {
    connection = c;
  }

  /// <summary>
  /// Make this your username, which should be provided.
  /// </summary>
  /// <returns>Returns the username of the player.</returns>
  public abstract String username();

  /// <summary>
  /// Make this your password, which should be provided.
  /// </summary>
  /// <returns>Returns the password of the player.</returns>
  public abstract String password();

  /// <summary>
  /// This is run once on turn one before run().
  /// </summary>
  public abstract void init();

  /// <summary>
  /// This is run every turn.
  /// </summary>
  /// <returns>
  /// Return true to end turn, false to resynchronize with the 
  /// server and run again.
  /// </returns>
  public abstract bool run();

  /// <summary>
  /// This is run once after your last turn.
  /// </summary>
  public abstract void end();

  /// <summary>
  /// Synchronizes with the server, then calls run().
  /// </summary>
  /// <returns>
  /// Return true to end turn, false to resynchronize with the 
  /// server and run again.
  /// </returns>
  public bool startTurn()
  {
    int count = 0;
    iteration++;

    count = Client.getPlayerCount(connection);
    players = new Player[count];
    for(int i = 0; i < count; i++)
      players[i] = new Player(Client.getPlayer(connection, i));

    count = Client.getMappableCount(connection);
    mappables = new Mappable[count];
    for(int i = 0; i < count; i++)
      mappables[i] = new Mappable(Client.getMappable(connection, i));

    count = Client.getPlantCount(connection);
    plants = new Plant[count];
    for(int i = 0; i < count; i++)
      plants[i] = new Plant(Client.getPlant(connection, i));

    count = Client.getMutationCount(connection);
    mutations = new Mutation[count];
    for(int i = 0; i < count; i++)
      mutations[i] = new Mutation(Client.getMutation(connection, i));

    if(!initialized)
    {
      initialized = true;
      init();
    }

    return run();
  }

  /// <summary>
  /// The width of the total map.
  /// </summary>
  /// <returns>Returns the width of the total map.</returns>
  public int mapWidth()
  {
    int value = Client.getMapWidth(connection);
    return value;
  }

  /// <summary>
  /// The height of the total map.
  /// </summary>
  /// <returns>Returns the height of the total map.</returns>
  public int mapHeight()
  {
    int value = Client.getMapHeight(connection);
    return value;
  }

  /// <summary>
  /// The current turn number.
  /// </summary>
  /// <returns>Returns the current turn number.</returns>
  public int turnNumber()
  {
    int value = Client.getTurnNumber(connection);
    return value;
  }

  /// <summary>
  /// The maximum number of Plants allowed per player.
  /// </summary>
  /// <returns>Returns the maximum number of Plants allowed per player.</returns>
  public int maxPlants()
  {
    int value = Client.getMaxPlants(connection);
    return value;
  }

  /// <summary>
  /// The id of the current player.
  /// </summary>
  /// <returns>Returns the id of the current player.</returns>
  public int playerID()
  {
    int value = Client.getPlayerID(connection);
    return value;
  }

  /// <summary>
  /// What number game this is for the server.
  /// </summary>
  /// <returns>Returns what number game this is for the server.</returns>
  public int gameNumber()
  {
    int value = Client.getGameNumber(connection);
    return value;
  }

  /// <summary>
  /// The maximum distance that a bumbleweed can move during an uproot
  /// </summary>
  /// <returns>Returns the maximum distance that a bumbleweed can move during an uproot</returns>
  public int bumbleweedSpeed()
  {
    int value = Client.getBumbleweedSpeed(connection);
    return value;
  }

  /// <summary>
  /// The damage taken from standing in a pool
  /// </summary>
  /// <returns>Returns the damage taken from standing in a pool</returns>
  public int poolDamage()
  {
    int value = Client.getPoolDamage(connection);
    return value;
  }

  /// <summary>
  /// The strength buff taken from standing in a pool
  /// </summary>
  /// <returns>Returns the strength buff taken from standing in a pool</returns>
  public int poolBuff()
  {
    int value = Client.getPoolBuff(connection);
    return value;
  }

  /// <summary>
  /// The strength debuff caused by an enemy Titan
  /// </summary>
  /// <returns>Returns the strength debuff caused by an enemy Titan</returns>
  public int titanDebuff()
  {
    int value = Client.getTitanDebuff(connection);
    return value;
  }

  /// <summary>
  /// The rate which your spores return to maxSpores
  /// </summary>
  /// <returns>Returns the rate which your spores return to maxSpores</returns>
  public int sporeRate()
  {
    int value = Client.getSporeRate(connection);
    return value;
  }

  /// <summary>
  /// The maximum number of spores a player can have
  /// </summary>
  /// <returns>Returns the maximum number of spores a player can have</returns>
  public int maxSpores()
  {
    int value = Client.getMaxSpores(connection);
    return value;
  }

  /// <summary>
  /// The range of uproot
  /// </summary>
  /// <returns>Returns the range of uproot</returns>
  public int uprootRange()
  {
    int value = Client.getUprootRange(connection);
    return value;
  }
}
